core/src/main/scala/kafka/controller/ControllerRequestMerger.scala

```diff
@@ -0,0 +1,232 @@
+    } else {
+      controllerState = newControllerState
+    }
+
+    request match {
+      case leaderAndIsrRequest : LeaderAndIsrRequest.Builder => addLeaderAndIsrRequest(leaderAndIsrRequest, callback)
+      case updateMetadataRequest : UpdateMetadataRequest.Builder => addUpdateMetadataRequest(updateMetadataRequest)
+      case stopReplicaRequest: StopReplicaRequest.Builder => addStopReplicaRequest(stopReplicaRequest, callback)
+    }
+  }
+
+  def isLeaderAndIsrReplaceable(newState: LeaderAndIsrPartitionState, currentState: LeaderAndIsrPartitionState): Boolean = {
+    newState.brokerEpoch() > currentState.brokerEpoch() || newState.leaderEpoch() > currentState.leaderEpoch()
+  }
+
+  def mergeLeaderAndIsrPartitionState(incomingState: LeaderAndIsrPartitionState,
+    queuedStates: util.LinkedList[LeaderAndIsrPartitionState]): Unit = {
+    // keep merging requests from the tail of the queued States
+    while (!queuedStates.isEmpty && isLeaderAndIsrReplaceable(incomingState, queuedStates.getLast)) {
+      queuedStates.pollLast()
+    }
+    val inserted = queuedStates.offerLast(incomingState)
+    if (!inserted) {
+      throw new IllegalStateException(s"Unable to insert LeaderAndIsrPartitionState $incomingState to the merger queue")
+    }
+  }
+
+  private def addLeaderAndIsrRequest(request: LeaderAndIsrRequest.Builder,
+    callback: AbstractResponse => Unit): Unit = {
+    request.partitionStates().forEach{partitionState => {
+      val transformedPartitionState = LiCombinedControlTransformer.transformLeaderAndIsrPartition(partitionState, request.maxBrokerEpoch())
+
+      val topicPartition = new TopicPartition(partitionState.topicName(), partitionState.partitionIndex())
+      val queuedStates = leaderAndIsrPartitionStates.getOrElseUpdate(topicPartition,
+        new util.LinkedList[LeaderAndIsrPartitionState]())
+
+      mergeLeaderAndIsrPartitionState(transformedPartitionState, queuedStates)
+    }}
+    leaderAndIsrLiveLeaders = request.liveLeaders()
+    leaderAndIsrCallback = callback
+  }
+
+  private def addUpdateMetadataRequest(request: UpdateMetadataRequest.Builder): Unit = {
+    request.partitionStates().forEach{partitionState => {
+      val transformedPartitionState = LiCombinedControlTransformer.transformUpdateMetadataPartition(partitionState)
+
+      val topicPartition = new TopicPartition(partitionState.topicName(), partitionState.partitionIndex())
+
+      updateMetadataPartitionStates.put(topicPartition, transformedPartitionState)
+    }}
+
+    updateMetadataLiveBrokers.clear()
+    request.liveBrokers().forEach{liveBroker =>
+      updateMetadataLiveBrokers.add(LiCombinedControlTransformer.transformUpdateMetadataBroker(liveBroker))
+    }
+  }
+
+  def isStopReplicaReplaceable(newState: StopReplicaPartitionState, currentState: StopReplicaPartitionState): Boolean = {
+    newState.brokerEpoch() > currentState.brokerEpoch()
+  }
+
+  def mergeStopReplicaPartitionState(incomingState: StopReplicaPartitionState,
```

Comment:
how do you maintain the relative order of stop replica requests and leaderAndISR requests that are combined together.

Reply:
I discussed with Wesley offline, and we decided to bring back the cancellation logic. Here is our rationale:

> For the case LeaderAndIsr following StopReplica (delete=true), a) if the LeaderAndIsr is become-follower, I think it doesn’t matter whether the deletion happens since the truncation will take care of the log convergence b) if the LeaderAndIsr is become-leader, there are the cases b1) due to topic deletion, we shouldn’t worry since this should only happen after the previous deletion has succeeded b2) due to partition reassignment, e.g. a removed replica suddenly needs to serve as the leader, e.g. due to an aborted partition reassignment. This seems impossible without the broker first becoming a follower, join the ISR.
> On the high level, the thinking is this: the controller should only tell a broker to become a leader when it’s safe for it to become the leader, that is the controller must always make sure the conditions are correct before sending a become-leader request to the broker. Along this line of thinking, then the cancellation logic would make sense.